Trying to call Beehive.addEntity with a BlockPopulator leads to IllegalStateException

### Expected behavior

The following code should spawn beehives with bees in them during block population
```java
	@Override
	public void populate(WorldInfo worldInfo, Random random, int chunkX, int chunkZ, LimitedRegion lr) {
		int targetX = chunkX*16 - 8;
		int targetZ = chunkZ*16 - 8;
		assert lr.isInRegion(targetX, 51, targetZ);

		lr.setType(targetX,51,targetZ,Material.BEE_NEST);
		Beehive bukkitBeehive = (Beehive) lr.getBlockState(targetX, 51, targetZ);
		Bee bee = (Bee) lr.createEntity(new Location(bukkitBeehive.getWorld(),targetX, 55, targetZ), Bee.class);
		bukkitBeehive.addEntity(bee);
	}
```

### Observed/Actual behavior

The code with a block populator will cause an IllegalStateException as Beehive.addEntity tries to call EntityRemoveEvent
```
Caused by: java.lang.IllegalStateException: EntityRemoveEvent may only be triggered synchronously.
        at io.papermc.paper.plugin.manager.PaperEventManager.callEvent(PaperEventManager.java:42) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at io.papermc.paper.plugin.manager.PaperPluginManagerImpl.callEvent(PaperPluginManagerImpl.java:131) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:628) ~[paper-api-1.21.5-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.craftbukkit.event.CraftEventFactory.callEntityRemoveEvent(CraftEventFactory.java:1964) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at net.minecraft.world.entity.Entity.setRemoved(Entity.java:4886) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at net.minecraft.world.entity.Entity.remove(Entity.java:687) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at net.minecraft.world.entity.LivingEntity.remove(LivingEntity.java:726) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at net.minecraft.world.entity.Entity.discard(Entity.java:619) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at net.minecraft.world.level.block.entity.BeehiveBlockEntity.addOccupant(BeehiveBlockEntity.java:204) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at org.bukkit.craftbukkit.block.CraftBeehive.addEntity(CraftBeehive.java:86) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at org.bukkit.craftbukkit.block.CraftBeehive.addEntity(CraftBeehive.java:17) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at SpawnerBlockstateBug.jar/org.worldgentutorial.main.WorldGenTutorialBlockPopulator.populate(WorldGenTutorialBlockPopulator.java:27) ~[SpawnerBlockstateBug.jar:?]
        at net.minecraft.world.level.chunk.ChunkGenerator.applyBiomeDecoration(ChunkGenerator.java:456) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at org.bukkit.craftbukkit.generator.CustomChunkGenerator.applyBiomeDecoration(CustomChunkGenerator.java:300) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at net.minecraft.world.level.chunk.status.ChunkStatusTasks.generateFeatures(ChunkStatusTasks.java:145) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at net.minecraft.world.level.chunk.status.ChunkStep.apply(ChunkStep.java:66) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        at ca.spottedleaf.moonrise.patches.chunk_system.scheduling.task.ChunkUpgradeGenericStatusTask.run(ChunkUpgradeGenericStatusTask.java:99) ~[paper-1.21.5.jar:1.21.5-5-8f62e0f]
        ... 7 more
```

### Steps/models to reproduce

Running the attached plugin with bukkit.yml's generator for the default world set to WorldGenTutorialPlugin
```yml
worlds:
  world:
    generator: WorldGenTutorial
```

### Plugin and Datapack List

Attached is a plugin (build with gradlew build) that triggers the bug by running the above code in blockpopulator. Don't mind the name, I've been reusing the same jar.

[SpawnerBlockstateBug.zip](https://github.com/user-attachments/files/19725064/SpawnerBlockstateBug.zip)

### Paper version

This server is running Paper version 1.21.5-5-main@8f62e0f (2025-04-13T09:54:28Z) (Implementing API version 1.21.5-R0.1-SNAPSHOT)

### Other

_No response_

--- a/paper-server/patches/sources/net/minecraft/server/level/ServerLevel.java.patch
+++ b/paper-server/patches/sources/net/minecraft/server/level/ServerLevel.java.patch
@@ -1065,10 +1065,11 @@
      @Override
      public CrashReportCategory fillReportDetails(CrashReport report) {
          CrashReportCategory crashReportCategory = super.fillReportDetails(report);
-@@ -1712,6 +_,7 @@
+@@ -1712,6 +_,8 @@
      final class EntityCallbacks implements LevelCallback<Entity> {
          @Override
          public void onCreated(Entity entity) {
++            entity.inWorld = true; // CraftBukkit - Mark entity as in world
 +            entity.setOldPosAndRot(); // Paper - update old pos / rot for new entities as it will default to Vec3.ZERO
          }
  
@@ -1108,11 +1109,10 @@
                      String string = "onTrackingStart called during navigation iteration";
                      Util.logAndPauseIfInIde(
                          "onTrackingStart called during navigation iteration", new IllegalStateException("onTrackingStart called during navigation iteration")
-@@ -1755,10 +_,52 @@
+@@ -1755,10 +_,51 @@
              }
  
              entity.updateDynamicGameEventListener(DynamicGameEventListener::add);
-+            entity.inWorld = true; // CraftBukkit - Mark entity as in world
 +            entity.valid = true; // CraftBukkit
 +            ServerLevel.this.getChunkSource().addEntity(entity); // Paper - ignore and warn about illegal addEntity calls instead of crashing server
 +            // Paper start - Entity origin API

--- a/paper-server/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/paper-server/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1961,6 +1961,11 @@ public class CraftEventFactory {
             return;
         }
 
+        // SPIGOT-8041: Do not call event unless entity has been spawned into world
+        if (!entity.inWorld) {
+            return;
+        }
+
         Bukkit.getPluginManager().callEvent(new EntityRemoveEvent(entity.getBukkitEntity(), cause));
     }
 

